### PR TITLE
Add host side on device test layers for nodes

### LINKED ISSUE #41: Add Host-Side + On-Device Test Layers for Nodes
## Description
This issue tracks introducing a two-layer test strategy for embedded nodes to improve reliability without slowing down development velocity.

Host-side tests will validate pure logic quickly on desktop CI using `#[test]` (and optionally property-based tests like `quickcheck`).  
On-device tests will validate hardware/runtime behavior using `embedded-test` on real target hardware.

## Proposed Changes
- Add a host test suite (`#[cfg(test)]`, `#[test]`) for pure logic modules.
- Add optional property-based tests (`quickcheck`) for edge-heavy logic.
- Add an on-device `embedded-test` suite for hardware/runtime integration points.
- Define a test matrix by layer:
  - Host: parsers, retry/streak logic, state transitions, payload/build logic, config behavior.
  - Device: UART/modem init flow, connectivity lifecycle, transport control transitions, recovery behavior.
- Add CI jobs:
  - host tests on every PR,
  - device tests on dedicated runner/manual gate.
- Add short testing docs (`how to run host tests`, `how to run device tests`, expected environment).

## Acceptance Criteria
- [ ] Host test target exists and runs with `cargo test` in CI.
- [ ] Critical logic paths are covered by host tests (recovery stages, retry behavior, serialization/parsing).
- [ ] On-device `embedded-test` target exists and can be run reproducibly on hardware.
- [ ] At least one end-to-end modem lifecycle test is validated on device (init -> connect -> upload -> recover).
- [ ] Test execution instructions are documented for both host and device paths.
- [ ] PR template/definition of done references both test layers where applicable.

### PR DESCRIPTION
**Pull request by FrancescoLanthaler**
## Summary
This PR introduces a two-layer test strategy for the ESP32 edge node: fast host-side logic tests in CI and reproducible local device lifecycle tests on real hardware.

### Main Changes:
- Added a dedicated host test target at `tests/host` with coverage for parser behavior, retry/recovery logic, state/policy transitions, and ingest payload/config behavior.
- Added property-based tests (`quickcheck`) to the host suite and made them mandatory in CI.
- Added a local ESP32 device test target at `tests/device` with serial lifecycle validation (`init -> connect -> upload -> recover`) via `run-modem-lifecycle.sh`.
- Added GitHub Actions host test workflow (`pull_request` only) running:
  - `cargo test --manifest-path tests/host/Cargo.toml --all-targets --features property-tests`
- Updated PR template validation checklist to require both host and local device validation where applicable.

### Fixes:
- Fixed device test runner robustness for local environments (fallback when `rg` is missing, shell issues around optional port args, clearer PASS/FAIL handling).
- Aligned CI/docs wording with protected-`main` workflow (PR-triggered host CI, device tests local-only).

### Pure Functions:
- Added small pure policy helpers for testability:
  - GSM recovery stage helpers (`connect_recovery_stage_index`, `upload_recovery_stage_index`)
  - Orchestration policy helpers (`bump_counter`, `should_fallback_to_saving`, `should_enter_sleep`)

### Device tests (local ESP32):
- Command: cd tests/device && ./run-modem-lifecycle.sh
- Key log lines:
  - INFO - HTTPACTION result: method=1, status=200, len=29
  - INFO - DEVICE_TEST_PASS modem_lifecycle_serial
  - Device lifecycle test passed.

### CODE REVIEW

File: tests/host/tests/logic_suite.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+        fn deduplicate_output_is_bounded(values: Vec<u16>) -> bool {
+            probe_counter::deduplicate_probes(&values) as usize <= values.len()
+        }
+    }
+}
+
+#[test]
+fn ingest_url_builds_expected_suffix() {
+    let url = config::ingest_url().expect("ingest URL should be buildable");
+    assert_eq!(url.as_str(), "https://api.trailsense.daugt.com/ingest");
```
[the remark this answers is not in the document]
Comment (FrancescoLanthaler) [reply tests/host/tests/logic_suite.rs:56]: We should keep it hardcoded since the URL should never change if possible, to avoid nodes in the wild breaking.

If we update the domain before deploying nodes we just need to update the test.
